lib/safe-http-client.js

```diff
@@ -0,0 +1,134 @@
+import dns from 'node:dns/promises'
+
+import {fetch} from 'undici'
+import ipaddr from 'ipaddr.js'
+
+import {IMAGE_MIME_TYPES} from './constants.js'
+
+export class HttpError extends Error {
+  /**
+   * @param {number} statusCode
+   * @param {string} message
+   */
+  constructor(statusCode, message) {
+    super(message)
+    this.statusCode = statusCode
+    this.name = 'HttpError'
+
+    // Captures the stack trace, excluding the constructor call from the stack trace.
+    if (Error.captureStackTrace) {
+      Error.captureStackTrace(this, HttpError)
+    }
+  }
+}
+
+export class SafeHttpClient {
+  /** @param {number} [maxSize] */
+  constructor(maxSize) {
+    this.maxSize = maxSize
+  }
+
+  /**
+   * Check if the URL is a valid URL or IP, that the prototol is valid,
+   * and the host is a safe unicast address.
+   * @param {string} url
+   */
+  static async checkUrl(url) {
+    // Throws if the URL is invalid
+    const validUrl = new URL(url)
+    const {protocol, hostname} = validUrl
```

Comment:
```suggestion
  /**
   * Check that the URL is valid, the prototol is allowed, and that the host is
   * a safe unicast address.
   *
   * @param {string} url
   *   URL to check.
   * @returns {Promise<URL>}
   *   URL object.
   */
  static async checkUrl(url) {
    const validUrl = new URL(url)
    const {hostname, protocol} = validUrl
```

Docs, sorting.

Q: when a function is called `checkUrl`, I‘d assume it would only *check* and return a boolean, instead of returning a parsed URL. Throwing errors in “check” also seems not intuitive to me.

Alternatively, I can understand a `checkUrl` that throws errors and returns `Promise<undefined>`, aka nothing.

Last, it might make sense to use different status codes for different errors? Simplest would perhaps be to inline it in `server.js`. Or throw `HttpError`s and use the `error.statusCode || 400` on them in `server.js`?

Quickly thinking, maybe:
invalid url -> 400
bad protocol -> 403
bad url host (1, dns.lookup) -> 400
bad url host (2, ipaddr) -> 403

Reply:
> Throwing errors in “check” also seems not intuitive to me

For me it's the opposite, I do expect errors in this function. 

> Alternatively, I can understand a `checkUrl` that throws errors and returns `Promise<undefined>`, aka nothing.

That seems fine yes!

>  Last, it might make sense to use different status codes for different errors?

I don't think so, they are all client-side mistakes (intentional or not). I'm also happier not exposing too many details of the internals and rejecting everything "bad" regarding URLs with `400`.